core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala

```diff
@@ -139,47 +140,51 @@ class DownsampledTimeSeriesShard(rawDatasetRef: DatasetRef,
 
   private def hour(millis: Long = System.currentTimeMillis()) = millis / 1000 / 60 / 60
 
-  def recoverIndex(): Future[Unit] = {
+  def recoverIndex(): Future[Long] = {
     if (downsampleConfig.enablePersistentIndexing) {
       partKeyIndex.getCurrentIndexState() match {
         case (IndexState.Empty, _)              =>
           logger.info("Found index state empty, bootstrapping downsample index")
-          recoverIndexInternal
+          recoverIndexInternal(None)
+        case (IndexState.Synced, checkpointMillis)             =>
+          logger.info("Found index state empty, bootstrapping downsample index")
+          recoverIndexInternal(checkpointMillis)
         case _                                  =>
           // FIXME: Part ids in case of indexRefresh are assumed to monotonically increase, however, in case recovery
           //  is done from an existing timestamp, we need a way not only to store the state of the ids recycled but
           //  latest allocated partId to avoid assigning the same id again.
-          logger.info(s"Nothing to recover the index for dataset=$indexDataset shard=$shardNum" +
-                      s" starting index refresh thread")
-          indexRefresh().runToFuture(housekeepingSched)
+          //logger.info(s"Nothing to recover the index for dataset=$indexDataset shard=$shardNum" +
+          //            s" starting index refresh thread")
+          //indexRefresh().runToFuture(housekeepingSched)
+          ???
       }
     } else {
       // Index persistence is not enabled, this will simply follow the path for existing index recovery
-      recoverIndexInternal
+      recoverIndexInternal(None)
     }
   }
 
-  private def recoverIndexInternal: Future[Unit] = {
+  private def recoverIndexInternal(checkpointMillis: Option[Long]): Future[Long] = {
```

Comment:
Ideally we use refresh to update the index since last sync time and thus recover, which is called from bootstrap will be called only if full rebuild is needed. Check out the implementation of ``createPartitionID`` it simply increments a counter where as increment update of index can not  use an incrementing counter for partId as we had discussed. Given this information  of ``createPartitionID``, ``recoverIndexInternal``  can only rebuild the index from scratch and not from any given timestamp and ``refreshWithDownsamplePartKeys`` will be the way to do that

Reply:
so, I put a comment in the method, essentially telling that currently in production settings one should call the method only with None.